[SPARK-44034][INFRA] Add a new test group for sql module

LuciferYang opened the pull request:
### What changes were proposed in this pull request?
The purpose of this pr is to add a new test tag `SlowSQLTest` to the sql module, and identified some Suites with test cases more than 3 seconds, and apply it to GA testing task to reduce the testing pressure of the `sql others` group.

For branches-3.3 and branches-3.4, a tag that will not appear in the sql module was assigned to the new test group to avoid `java.lang.ClassNotFoundException` and make this group build only without running any tests.

### Why are the changes needed?
For a long time, the sql module UTs has only two groups: `slow` and `others`. The test cases in group `slow` are fixed, while the number of test cases in group `others` continues to increase, which has had a certain impact on the testing duration and stability of group `others`. 

So this PR proposes to add a new testing group to share the testing pressure of `sql others` group, which has made the testing time of the three groups more average, and hope it can improve the stability of the GA task.




### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
Should monitor GA

LuciferYang committed:
add new test group

LuciferYang committed:
fix compile

LuciferYang commented on .github/workflows/build_and_test.yml:
```diff
@@ -180,7 +180,13 @@ jobs:
+            java: ${{ inputs.java }}
+            hadoop: ${{ inputs.hadoop }}
+            hive: hive2.3
+            excluded-tags: org.apache.spark.tags.ExtendedSQLTest,org.apache.spark.tags.ExtendedSQLCommandTest
```
Need to solve the issue of branch-3.4 without `ExtendedSQLCommandTest` ...

LuciferYang committed:
use expression

LuciferYang committed:
for ga

LuciferYang committed:
Merge branch 'upmaster' into new-sql-test-group

LuciferYang committed:
add more to this group

LuciferYang commented on common/tags/src/test/java/org/apache/spark/tags/ExtendedSQLCommandTest.java:
```diff
@@ -0,0 +1,30 @@
+@TagAnnotation
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD, ElementType.TYPE})
+public @interface ExtendedSQLCommandTest { }
```
Do you have a more recommended name? So that we can add more cases to this group

LuciferYang replied:
@zhengruifeng 
1.  Do you have any suggestions for the name of the tag?
2.  I am thinking of moving what type of test to this new group. One way is to have the same type of feature, such as now moving `Command` related cases to this group, and the other way is to move the first x% (possibly 30%?) of the slowest running cases to this group. Which do you think is more reasonable? Do you have any better suggestions?

zhengruifeng replied:
> Do you have any suggestions for the name of the new tag?

I feel `ExtendedSQLCommandTest` suggest only sql commands related tests can be used with this tag. 


> I am thinking of moving what type of test to this new group. One way is to have the same type of feature, such as now moving Command related cases to this group, and the other way is to move the first x% (possibly 30%?) of the slowest running cases to this group. Which do you think is more reasonable? Do you have any better suggestions?

I think we can just find the slowest tests, don't need to be the same group.

also cc @HyukjinKwon @dongjoon-hyun

LuciferYang replied:
Yes, I'm thinking of a more universal name, but I don't have any good ideas yet

LuciferYang committed:
fix compile

LuciferYang committed:
fix compile

LuciferYang committed:
Merge branch 'apache:master' into new-sql-test-group

LuciferYang committed:
for ga

zhengruifeng commented:
It make sense to split `sql - other` which is super slow.

btw, I am just wondering is it possible to fail a time-out (maybe 10mins?) test suite and continue remaining tests?

LuciferYang commented:
> It make sense to split `sql - other` which is super slow.
> 
> btw, I am just wondering is it possible to fail a time-out (maybe 10mins?) test suite and continue remaining tests?

This needs to be investigated, and I am not sure if scalatest has a timeout feature similar to Junit

LuciferYang committed:
add ExtendedStructuredStreamingTest

LuciferYang committed:
remove tag of V1WriteCommandSuiteBase

LuciferYang commented:
> It make sense to split `sql - other` which is super slow.
> 
> btw, I am just wondering is it possible to fail a time-out (maybe 10mins?) test suite and continue remaining tests?

@zhengruifeng  perhaps we can use `failAfter` to wrap the` testBody` to control the timeout: https://github.com/apache/spark/pull/41590

LuciferYang committed:
Merge branch 'apache:master' into new-sql-test-group

LuciferYang committed:
for ga

LuciferYang committed:
make slow tag

LuciferYang commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/DeprecatedStreamingAggregationSuite.scala:
```diff
@@ -24,7 +24,9 @@ import org.apache.spark.sql.execution.streaming.state.StreamingAggregationStateM
 import org.apache.spark.sql.streaming.OutputMode._
+import org.apache.spark.tags.SlowSQLTest
 
+@SlowSQLTest
```
Some test in streaming dir is not slow, need more check

LuciferYang committed:
rename to SQLTestGroupOne

LuciferYang commented:
@dongjoon-hyun Should we continue trying this solution?

LuciferYang committed:
Merge branch 'upmaster' into new-sql-test-group

dongjoon-hyun commented:
I still support the idea to split `sql - others` into two groups, @LuciferYang .

LuciferYang commented:
> I still support the idea to split `sql - others` into two groups, @LuciferYang .

Thanks ~

dongjoon-hyun commented on common/tags/src/test/java/org/apache/spark/tags/SQLTestGroupOne.java:
```diff
@@ -0,0 +1,30 @@
+@TagAnnotation
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD, ElementType.TYPE})
+public @interface SQLTestGroupOne { }
```
May I ask why we choose this name, `SQLTestGroupOne`?

So, this PR proposes to have the following group name?
- (None) Default
- ExtendedSQLTest
- SQLTestGroupOne

LuciferYang replied:
This is just because I didn't come up with a more suitable name, perhaps we could call it `SlowSQLTest`. Currently, most of the suites in this group contain test cases that run for more than 3 seconds (further screening is needed and some not slow ones need to be removed)

dongjoon-hyun replied:
Do you think about splitting by module which is more intuitive?

dongjoon-hyun replied:
According to this PR, all test suites under `package org.apache.spark.sql.streaming` seems to be in `SQLTestGroupOne`, doesn't it?

LuciferYang replied:
> According to this PR, all test suites under `package org.apache.spark.sql.streaming` seems to be in `SQLTestGroupOne`, doesn't it?

No, just slower part with `SlowSQLTest `, for example, `ContinuousQueuedDataReaderSuite`, `ContinuousEpochBacklogSuite`, `LongOffsetSuite`, etc. are not with `SlowSQLTest `.

LuciferYang replied:
> Do you think about splitting by module which is more intuitive?

I think it would be more intuitive. 
The advantage is that the grouping way is relatively flexible, which is conducive to further  group splitting in the future. The disadvantage is that multiple tags need to be added according to module and and the number of tags may continue to increase.

LuciferYang committed:
remove not slow

LuciferYang committed:
rename to SlowTest

LuciferYang committed:
Merge branch 'upmaster' into new-sql-test-group

LuciferYang commented:
run `build/sbt clean "sql/test" -Dtest.exclude.tags=org.apache.spark.tags.ExtendedSQLTest,org.apache.spark.tags.SlowSQLTest`

QueryExecutionSuite will failed.

```
2023-06-15T19:58:34.4105460Z [0m[[0m[0minfo[0m] [0m[0m[32mQueryExecutionSuite:[0m[0m
2023-06-15T19:58:34.5395268Z [0m[[0m[0minfo[0m] [0m[0m[32m- dumping query execution info to a file (77 milliseconds)[0m[0m
2023-06-15T19:58:34.5856902Z [0m[[0m[0minfo[0m] [0m[0m[32m- dumping query execution info to an existing file (49 milliseconds)[0m[0m
2023-06-15T19:58:34.6099849Z [0m[[0m[0minfo[0m] [0m[0m[32m- dumping query execution info to non-existing folder (25 milliseconds)[0m[0m
2023-06-15T19:58:34.6136467Z [0m[[0m[0minfo[0m] [0m[0m[32m- dumping query execution info by invalid path (4 milliseconds)[0m[0m
2023-06-15T19:58:34.6425071Z [0m[[0m[0minfo[0m] [0m[0m[32m- dumping query execution info to a file - explainMode=formatted (28 milliseconds)[0m[0m
2023-06-15T19:58:34.7084916Z [0m[[0m[0minfo[0m] [0m[0m[32m- limit number of fields by sql config (66 milliseconds)[0m[0m
2023-06-15T19:58:34.7432299Z [0m[[0m[0minfo[0m] [0m[0m[32m- check maximum fields restriction (34 milliseconds)[0m[0m
2023-06-15T19:58:34.7554546Z [0m[[0m[0minfo[0m] [0m[0m[32m- toString() exception/error handling (11 milliseconds)[0m[0m
2023-06-15T19:58:34.7621424Z [0m[[0m[0minfo[0m] [0m[0m[32m- SPARK-28346: clone the query plan between different stages (6 milliseconds)[0m[0m
2023-06-15T19:58:34.8001412Z [0m[[0m[0minfo[0m] [0m[0m[31m- Logging plan changes for execution *** FAILED *** (12 milliseconds)[0m[0m
2023-06-15T19:58:34.8007977Z [0m[[0m[0minfo[0m] [0m[0m[31m  testAppender.loggingEvents.exists(((x$10: org.apache.logging.log4j.core.LogEvent) => x$10.getMessage().getFormattedMessage().contains(expectedMsg))) was false (QueryExecutionSuite.scala:232)[0m[0m
```

but furn `build/sbt "sql/testOnly *QueryExecutionSuite"` not this issue, need to investigate.

LuciferYang committed:
ignore Logging plan changes for execution

LuciferYang commented on sql/core/src/test/scala/org/apache/spark/sql/execution/QueryExecutionSuite.scala:
```diff
@@ -220,7 +220,8 @@ class QueryExecutionSuite extends SharedSparkSession {
 
-  test("Logging plan changes for execution") {
+  // TODO(SPARK-44074): re-enable this test after SPARK-44074 resolved
+  ignore("Logging plan changes for execution") {
```
https://github.com/apache/spark/pull/41552#issuecomment-1593887715

Due to this pr move some test suites out of `others` group,


run `build/sbt clean "sql/test" -Dtest.exclude.tags=org.apache.spark.tags.ExtendedSQLTest,org.apache.spark.tags.SlowSQLTest`

`Logging plan changes for execution` will test failed.


but run `run build/sbt "sql/testOnly *QueryExecutionSuite"` no failed.


And this issue can be reproduced locally, created SPARK-44074 to track this issue and ignore this test in the current pr.

LuciferYang committed:
add more slow

LuciferYang committed:
add more slow

LuciferYang committed:
add more tests

LuciferYang commented:
Now 

- sql - extended tests(The original slow tests): 1h 27m 25s
- sql - slow tests: 1h 20m 18s
- sql - others tests: 1h 18m 43s

The testing time for the three sql test tasks has been relatively average :)

LuciferYang committed:
Merge branch 'apache:master' into new-sql-test-group

beliefer committed:
[SPARK-43928][SQL][PYTHON][CONNECT] Add bit operations to Scala, Python and Connect API

### What changes were proposed in this pull request?
This PR want add bit operations to Scala, Python and Connect API.
These API show below.

- bit_and
- bit_count
- bit_get
- bit_or
- bit_xor
- getbit

### Why are the changes needed?
Add bit operations to Scala, Python and Connect API

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
New test cases.

Closes #41608 from beliefer/SPARK-43928.

Authored-by: Jiaan Geng <beliefer@163.com>
Signed-off-by: Ruifeng Zheng <ruifengz@apache.org>

Commit:
[MINOR][DOCS] Remove double quote to fix link

### What changes were proposed in this pull request?

typo leading to broken link in docs

### Why are the changes needed?

fix broken link in docs

### Does this PR introduce _any_ user-facing change?

Yes

### How was this patch tested?

clicked on fixed link

Closes #41595 from danielsparing/typo.

Authored-by: Daniel Sparing <daniel.sparing@databricks.com>
Signed-off-by: Ruifeng Zheng <ruifengz@apache.org>

itholic committed:
[SPARK-44052][CONNECT][PS] Add util to get proper Column or DataFrame class for Spark Connect

### What changes were proposed in this pull request?

This PR proposes to add util to get proper Column or DataFrame class for Spark Connect.

### Why are the changes needed?

To eliminate code duplication and obtain the appropriate Spark Connect Column and Spark Connect DataFrame classes, utility functions `get_column_class` and `get_dataframe_class` have been added.

### Does this PR introduce _any_ user-facing change?

No, it's code refactoring.

### How was this patch tested?

The existing CI should pass.

Closes #41570 from itholic/SPARK-43703.

Authored-by: itholic <haejoon.lee@databricks.com>
Signed-off-by: Ruifeng Zheng <ruifengz@apache.org>

wangyum committed:
[MINOR][K8S][DOCS] Fix all dead links for K8s doc

### What changes were proposed in this pull request?

This PR fixes all dead links for K8s doc.

### Why are the changes needed?

<img width="797" alt="image" src="https://github.com/apache/spark/assets/5399861/3ba3f048-776c-42e6-b455-86e90b6ef22f">

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manual test.

Closes #41635 from wangyum/kubernetes.

Authored-by: Yuming Wang <yumwang@ebay.com>
Signed-off-by: Yuming Wang <yumwang@ebay.com>

LuciferYang commented:
The commits of this pr is not clean, I have reopened a new one: https://github.com/apache/spark/pull/41638 @zhengruifeng @dongjoon-hyun @HyukjinKwon
